### PR TITLE
Drop OSX builds for python 3.9

### PR DESCRIPTION
**Pull request by fjetter**
Closes https://github.com/dask/distributed/issues/6040

I went a bit further than what has been discussed in the above issue, mostly to have things cleaned up a bit.

- I cleaned up the 3.9 build since I believe we should have one build with as minimal dependencies as possible (e.g. to verify that we don't have obscure import errors)
- We need to test all of our optional dependencies, this is now done in 3.9
- Lastly, we want to test against upstream changes for some of our important dependencies where we're testing against git tips

Based on this cleanup, I figured the most important build to keep around for OSX would be the "all dependencies" one which is 3.9 atm.

If this cleanup is too aggressive, we can go with [Matt's proposal](https://github.com/dask/distributed/issues/6040#issuecomment-1085848863) and just drop py3.9

### CODE REVIEW

**Comment by crusaderky** [comment]
I don't think this is wise. You're skipping all code specific to a Python version that test an optional dependency.

Artificial, but realistic example:

mymod.py:
```python
from __future__ import annotations
from typing import cast

def f(x: int | dict[str, int]):
    import scipy
    if some_condition:  # anything but isinstance()
        return scipy.foo(cast(int, x))
    else:
        return scipy.bar(cast(dict[str, int], x))
```
test_mymod.py:
```python
def test_f():
    pytest.import_or_skip("scipy")
    assert f(1) == ...
    assert f({"x": 1}) == ...
```
The above fails on Python 3.8 because of the ``cast(dict[str, int], x)`` without quotes; however the test suite won't catch it anymore!

**Comment by crusaderky** [comment]
Currently we have:

Python version | OS | Which deps | deps version
--- | --- | --- | ---
3.8 | Linux, MacOSX, Windows | mandatory, optional, Cython | latest
3.9 | Linux, MacOSX, Windows | mandatory, optional, CUDA | latest
3.10 | Linux, MacOSX, Windows | mandatory, optional | latest+git tip


I think we should change it to:

Python version | OS | Which deps | deps version
--- | --- | --- | ---
3.8 | Linux | mandatory, optional, CUDA | pinned to minimum
3.8 | Linux | mandatory only (no numpy!) | latest
3.8 | Linux, MacOSX, Windows | mandatory, optional, CUDA | latest
3.9 | Linux | mandatory, optional, CUDA | latest
3.10 | Linux, MacOSX, Windows | mandatory, optional, CUDA | latest+git tip

That's 9 workflows before and after, but with one less Mac workflow and *a lot* of added value.
I expect the first two workflows to require a lot of effort before they become green.

**Comment by fjetter** [comment]
> That's 9 workflows before and after, but with one less Mac workflow and a lot of added value.
I expect the first two workflows to require a lot of effort before they become green.

I'm open to setting it up this way but that requires much more effort than what I'm willing to put in this at this point in time.

What I wanted to rectify here is that we have "partial optional" all over the place but if there is more demand for a conversation, how about the following


Python version | OS | Which deps | deps version
--- | --- | --- | ---
3.8 | Linux, MacOSX, Windows | mandatory, optional-as-is, Cython, CUDA | latest
3.9 | Linux, Windows | mandatory, optional-as-is | latest
3.10 | Linux, Windows | mandatory, optional-as-is | latest+git tip


We have a few only-tested-here in the py3.9 build but I don't think there is a lot of sensitivity to OSX in there

**Comment by fjetter** [comment]
> I would not disable MacOS testing on 3.10 and I would not skip the compression libs either.

I do consider the compression libs and a few other things like scipy optional. Essentially everything that's currently not in requirements.txt

However, my suggestion in https://github.com/dask/distributed/pull/6073#issuecomment-1091830274 meant simply to move `pynvml` to py3.8 and drop the py3.9 build

> I would not disable MacOS testing on 3.10

why?

**Comment by crusaderky** [comment]
> > I would not disable MacOS testing on 3.10
> 
> why?

In 3.10 specifically, nothing comes to mind. But there have been cases in the past for a major OS-specific change. For example, in 3.8 the default mp-context changed from fork to spawn on MacOSX. It would be very bad practice to skip testing it just because nothing catches our eyes from the release notes.

**Comment by fjetter** [comment]
> In 3.10 specifically, nothing comes to mind. But there have been cases in the past for a major OS-specific change. For example, in 3.8 the default mp-context changed from fork to spawn on MacOSX. It would be very bad practice to skip testing it just because nothing catches our eyes from the release notes.

My argument is mostly that the cost is very high to keep this up and the likelihood for an error very small.

**Comment by fjetter** [comment]
Unrelated question: Do we need CUDA tests everywhere? Isn't this covered by the gpuCI?

**Comment by fjetter** [comment]
I opened https://github.com/dask/distributed/issues/6085 for any follow up conversations

File: continuous_integration/environment-3.9.yaml [hunk 1]
```diff
@@ -18,15 +18,13 @@ dependencies:
   - ipykernel
   - ipywidgets
   - jinja2
-  - joblib
```
Comment (fjetter) [inline continuous_integration/environment-3.9.yaml:21]: Last reference to this was removed in https://github.com/dask/distributed/pull/3040

**Pull request merged**